### ISSUE TITLE
DBRefs resulting in Compile Error

### ISSUE THREAD

**Issue by omnidan** [doctrine/mongodb-odm#1041]
I migrated some data and changed DBRefs in MongoDB which according to a guy from #mongodb is fine, and I compared the data to when I create a new object and it's the same.

When I create a new object, everything works fine, but when I migrate existing objects to have the same references as the new object (I'm reversing a bidirectional reference), it gives me a compile error and tells me that a proxy does not exist, yes the file does not exist but that's because the class it's trying to use is an abstract class, but has a discriminator map. It seems like a bug in the discriminator map when resolving @ReferenceMany in doctrine odm. It seems to always resolve to targetDocument regardless of the discriminatormap.

```Compile Error: Doctrine\Common\Proxy\AbstractProxyFactory::getProxyDefinition(): Failed opening required '/srv/from_git/api/app/cache/dev/doctrine/odm/mongodb/Proxies/__CG__TouchLayAuthBundleDocumentAuth.php' (include_path='.:/usr/share/php:/usr/share/pear')```

Why does it work with new objects though? This is kinda confusing - does anybody have a clue why this could be broken like that?

**Comment by malarzm** [comment]
@omnidan What kind of migration are you doing? This will be a blind guess, but if you haven't migrated discriminator value held in DBRef then ODM will not know what should be proxied. Anyway more details will be helpful :)

**Comment by omnidan** [comment]
@malarzm "The discriminator value held in DBRef" - How would I do this with mongo shell, I think that's exactly the issue. (I didn't find it when doing mongo shell find queries so I assumed everything I had was enough and doctrine would query by id and then find the class by using the discriminator field/map)

**Comment by malarzm** [comment]
Mongo shell by default doesn't show any additional stuff from DBRef, you had to [alter its prototype](http://stackoverflow.com/questions/14629701/mongo-dbref-additional-fields-are-invisible-in-mongoshell-how-to-display-them) to see it. I haven't tried altering DBRefs manually so googling it may be better :)

**Comment by omnidan** [comment]
@malarzm So what's the name of that field? Or even better, what query does doctrine run in mongodb to set this field?

**Comment by malarzm** [comment]
The name of the field depends on your configuration but default is ``_doctrine_class_name``. I suppose it can be just set in console with normal update however as I've said, I haven't been manipulating these by myself yet.. As for ODM, here's [method for creating DBRefs](https://github.com/doctrine/mongodb-odm/blob/master/lib/Doctrine/ODM/MongoDB/DocumentManager.php#L684) which you can utilize

**Comment by omnidan** [comment]
Thanks a lot! I used your links to add the `inspect` function to my mongo shell, then found out that it saves `"type": "product"`, type is my discriminator field and `"product"` is in the map.

So I added that manually to my references by setting `assigned_ids: {"$ref": "auths", "$id": id1, "type": "product"}` instead of `assigned_ids: DBRef("auths", id1)`

Now everything runs fine, thanks a lot!

**Comment by malarzm** [comment]
Glad it helped :)

**Comment by jmikola** [comment]
@omnidan: This is a long-standing problem with the MongoDB shell and, to a lesser degree, server. Doctrine's use of DBRef objects to store extra values goes back to some very early MongoDB versions, and this functionality is not documented (I'm not sure if it ever was, to be honest).

[SERVER-10777](https://jira.mongodb.org/browse/SERVER-10777) is probably the most relevant issue regarding the shell not displaying DBRef fields (including {{$db}}!). I've also caught inadvertent BC breaks in the server, such as [SERVER-12263](https://jira.mongodb.org/browse/SERVER-12263).

This personally bit me years ago during a data migration (documented [in this presentation](https://speakerdeck.com/jmikola/using-mongodb-responsibly?slide=31)). We would probably do well to talk about this in the ODM documentation, so I'll make a new issue to track that.

**Comment by omnidan** [comment]
@jmikola I agree, documenting this would save people a lot of time. It took me almost the whole day to figure out what's wrong with doctrine after the migration. A few hours in I did realize it might have to do with the discriminator fields but couldn't see the difference in the database (because neither mongolab, nor mongo shell display DBRefs correctly) - so I was like "that can't be it then". Seeing this presentation before starting the migration would've saved me so much time :P

Would be cool if mongo shell used the inspect function @malarzm posted for DBRefs (or at least implemented a function like that by default)
